### ISSUE TITLE
"Effects PURE and ML cannot be composed" when all.fst is not in scope

### ISSUE THREAD

**Issue by catalin-hritcu** [FStarLang/FStar#311]
There seems to be a problem with partially applied function definitions when `all.fst` is not in scope. I don't want to import `all.fst` in `micro-fstar.fst` (and proof files in general) because we only work with Pure things. But without `all.fst` partially applied functions start failing. For instance:
```
val esub_inc2 : var -> Tot exp
let esub_inc2 = esub_inc_gen 2
```
works well when `all.fst` (and all its dependencies: `set.fsi heap.fst st.fst`) is in scope, but not when they are removed:
```
micro-fstar.fst(303,0-303,30) : Error
Effects PURE and ML cannot be composed
```
Fully applying `esub_inc2` also works around the problem, so the following works even without `all.fst`:
```
let esub_inc2 x = esub_inc_gen 2 x
```

**Comment by catalin-hritcu** [comment]
Found a place later in the file for which eta expanding is not really an option, but instead need to add a silly unit argument to make this work:
```
type sub_neappears_all : s:sub -> x:var -> Type =
| SubNEAppearsAll : s:sub -> x:var ->
              ef: (y:var -> Lemma (not (eeappears x (Sub.es s y)))) ->
	      tf: (a:var -> Lemma (not (teappears x (Sub.ts s a)))) ->
	      sub_neappears_all s x

val einc_neappears_all : (sub_neappears_all sub_einc 0)
let einc_neappears_all =
  SubNEAppearsAll sub_einc 0
                  (fun y -> ())
                  (fun a -> ())
```

**Comment by nikswamy** [comment]
Ok, worth looking into in detail in the coming days ...

**Comment by catalin-hritcu** [comment]
It's actually worse than I thought. This problem happens even without partially applied functions, and it's hard to predict what works and what not:
```
// this works
val x : int
let x = 42

// this doesn't
val x' : y:int{y>0}
let x' = 42
```

**Comment by nikswamy** [comment]
Formally, we expect the effect lattice to always include Pure and All. Excluding All is not really part of the original design ... But i see that in your case it may be quite useful. Figuring out how to support it is will require a careful analysis of the code ...it relies on the ML effect being in scope pervasively. Our theory also is a bit unclear if the user defined monads are not required to be related to All in any way. (cf. Ghost is an exception )

**Comment by catalin-hritcu** [comment]
I don't think the universes solves this, but hopefully with the current DM4F work we will no longer assume a top of the effect "lattice" (just a bottom?), which should also fix this kind of issue.

**Comment by catalin-hritcu** [comment]
Actually `--universes` did fix the problem of this issue:
```
[hritcu@detained bug-reports]$ time fstar.exe --explicit_deps bug311.fst
/home/hritcu/Projects/fstar/pub/examples/bug-reports/bug311.fst(8,19-9,11) : Error
Effects PURE and ML cannot be composed
fstar.exe --explicit_deps bug311.fst  0.36s user 0.01s system 92% cpu 0.398 total
[hritcu@detained bug-reports]$ time fstar.exe --universes --explicit_deps bug311.fst
Verifying module: Bug311
All verification conditions discharged successfully
fstar.exe --universes --explicit_deps bug311.fst  0.49s user 0.01s system 94% cpu 0.535 total
```

**Comment by nikswamy** [comment]
Fyi, it's the autodeps feature enabled by universes that's responsible for fixing it.

Sent from a mobile device



On Sat, Jun 18, 2016 at 9:45 AM -0700, "Catalin Hritcu" <notifications@github.com<mailto:notifications@github.com>> wrote:


Closed #311<https://github.com/FStarLang/FStar/issues/311>.

-

**Comment by catalin-hritcu** [comment]
I had passed `--explicit_deps` though. It seems that only includes `prims.fst`.
```
[hritcu@detained bug-reports]$ fstar.exe --universes --explicit_deps --debug Low bug311.fst 
Checking module: Prims
/home/hritcu/Projects/fstar/pub/ulib/prims.fst(16,12-18,14) (Diagnostic): Positivity check is not yet implemented (Prims.l_False)
/home/hritcu/Projects/fstar/pub/ulib/prims.fst(18,14-22,5) (Diagnostic): Positivity check is not yet implemented (Prims.l_True)
/home/hritcu/Projects/fstar/pub/ulib/prims.fst(46,31-50,31) (Diagnostic): Positivity check is not yet implemented (Prims.c_and)
/home/hritcu/Projects/fstar/pub/ulib/prims.fst(53,42-58,25) (Diagnostic): Positivity check is not yet implemented (Prims.c_or)
/home/hritcu/Projects/fstar/pub/ulib/prims.fst(85,72-92,26) (Diagnostic): Positivity check is not yet implemented (Prims.dtuple2)
/home/hritcu/Projects/fstar/pub/ulib/prims.fst(190,50-194,38) (Diagnostic): Positivity check is not yet implemented (Prims.list)
/home/hritcu/Projects/fstar/pub/ulib/prims.fst(194,38-199,45) (Diagnostic): Positivity check is not yet implemented (Prims.pattern)
/home/hritcu/Projects/fstar/pub/ulib/prims.fst(210,33-215,26) (Diagnostic): Positivity check is not yet implemented (Prims.option)
/home/hritcu/Projects/fstar/pub/ulib/prims.fst(215,26-219,30) (Diagnostic): Positivity check is not yet implemented (Prims.either)
/home/hritcu/Projects/fstar/pub/ulib/prims.fst(219,30-224,32) (Diagnostic): Positivity check is not yet implemented (Prims.result)
/home/hritcu/Projects/fstar/pub/ulib/prims.fst(426,44-432,26) (Diagnostic): Positivity check is not yet implemented (Prims.tuple2)
/home/hritcu/Projects/fstar/pub/ulib/prims.fst(432,26-439,28) (Diagnostic): Positivity check is not yet implemented (Prims.tuple3)
/home/hritcu/Projects/fstar/pub/ulib/prims.fst(439,28-447,32) (Diagnostic): Positivity check is not yet implemented (Prims.tuple4)
/home/hritcu/Projects/fstar/pub/ulib/prims.fst(447,32-456,35) (Diagnostic): Positivity check is not yet implemented (Prims.tuple5)
/home/hritcu/Projects/fstar/pub/ulib/prims.fst(456,35-466,38) (Diagnostic): Positivity check is not yet implemented (Prims.tuple6)
/home/hritcu/Projects/fstar/pub/ulib/prims.fst(466,38-478,41) (Diagnostic): Positivity check is not yet implemented (Prims.tuple7)
/home/hritcu/Projects/fstar/pub/ulib/prims.fst(478,41-490,44) (Diagnostic): Positivity check is not yet implemented (Prims.tuple8)
/home/hritcu/Projects/fstar/pub/ulib/prims.fst(490,44-499,29) (Diagnostic): Positivity check is not yet implemented (Prims.dtuple3)
/home/hritcu/Projects/fstar/pub/ulib/prims.fst(499,29-510,29) (Diagnostic): Positivity check is not yet implemented (Prims.dtuple4)
Checking module: Bug311
Verifying module: Bug311
All verification conditions discharged successfully
```
